packages/subproviders/src/subproviders/trezor.ts

```diff
@@ -0,0 +1,160 @@
+
+const PRIVATE_KEY_PATH = `m/44'/60'/0'`;
+
+export class TrezorSubprovider extends BaseWalletSubprovider {
+    private readonly _publicKeyPath: string;
+    private _cachedAccounts: string[];
+    private readonly _trezorConnectClientApi: any;
+    /**
+     * Instantiates a TrezorSubprovider. Defaults to private key path set to `44'/60'/0'`.
+     * Must be initialized with trezor-connect API module https://github.com/trezor/connect.
+     * @param TrezorSubprovider config object containing trezor-connect API
+     * @return TrezorSubprovider instance
+     */
+    constructor(config: TrezorSubproviderConfig) {
+        super();
+        this._publicKeyPath = PRIVATE_KEY_PATH;
+        this._cachedAccounts = [];
+        this._trezorConnectClientApi = config.trezorConnectClientApi;
+    }
+    /**
+     * Retrieve a users Trezor account. The accounts are private key path derived, This method
+     * is automatically called when issuing a `eth_accounts` JSON RPC request via your providerEngine
+     * instance.
+     * @return An array of accounts
+     */
+    public async getAccountsAsync(): Promise<string[]> {
+        if (this._cachedAccounts.length) {
+            return this._cachedAccounts;
+        }
+        const accounts: string[] = [];
+        const response: TrezorConnectResponse = await this._trezorConnectClientApi.ethereumGetAddress({
+            path: this._publicKeyPath,
+            showOnTrezor: true,
+        });
+
+        if (response.success) {
+            const payload: TrezorGetAddressResponsePayload = response.payload;
+            accounts.push(payload.address);
+            this._cachedAccounts = accounts;
+        } else {
+            const payload: TrezorResponseErrorPayload = response.payload;
+            throw new Error(payload.error);
+        }
+
+        return accounts;
+    }
+    /**
+     * Signs a transaction on the Trezor with the account specificed by the `from` field in txParams.
+     * If you've added the TrezorSubprovider to your app's provider, you can simply send an `eth_sendTransaction`
+     * JSON RPC request, and this method will be called auto-magically. If you are not using this via a ProviderEngine
+     * instance, you can call it directly.
+     * @param txParams Parameters of the transaction to sign
+     * @return Signed transaction hex string
+     */
+    public async signTransactionAsync(txData: PartialTxParams): Promise<string> {
+        if (_.isUndefined(txData.from) || !addressUtils.isAddress(txData.from)) {
+            throw new Error(WalletSubproviderErrors.FromAddressMissingOrInvalid);
+        }
+        txData.value = txData.value ? txData.value : '0x0';
+        txData.data = txData.data ? txData.data : '0x';
+        txData.gas = txData.gas ? txData.gas : '0x0';
+        txData.gasPrice = txData.gasPrice ? txData.gasPrice : '0x0';
+
+        const accountIndex = this._cachedAccounts.indexOf(txData.from);
```

Comment:
This will be `-1` if they supply a `from` address that is not in `_cachedAccounts`. We should throw with a helpful error message in this case.

Reply:
Updated. Should either throw an `ADDRESS_NOT_FOUND` subprovider error for signing personal message and `FROM_ADDRESS_MISSING_OR_INVALID` when signing tx.